spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestStoragePartitionedJoins.java

```diff
@@ -0,0 +1,585 @@
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.spark.SparkSQLProperties;
+import org.apache.iceberg.spark.SparkSchemaUtil;
+import org.apache.iceberg.spark.SparkTestBaseWithCatalog;
+import org.apache.iceberg.spark.SparkWriteOptions;
+import org.apache.iceberg.spark.data.RandomData;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.analysis.NoSuchTableException;
+import org.apache.spark.sql.internal.SQLConf;
+import org.apache.spark.sql.types.StructType;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class TestStoragePartitionedJoins extends SparkTestBaseWithCatalog {
+
+  private static final String OTHER_TABLE_NAME = "other_table";
+
+  // open file cost and split size are set as 16 MB to produce a split per file
+  private static final String TABLE_PROPERTIES =
+      String.format(
+          "'%s' = 16777216, '%s' = 16777216",
+          TableProperties.SPLIT_SIZE, TableProperties.SPLIT_OPEN_FILE_COST);
+
+  private static final Map<String, String> SPJ_ON_SQL_CONF =
+      ImmutableMap.of(
+          SQLConf.V2_BUCKETING_ENABLED().key(),
+          "true",
+          SQLConf.REQUIRE_ALL_CLUSTER_KEYS_FOR_CO_PARTITION().key(),
+          "false",
+          SQLConf.ADAPTIVE_EXECUTION_ENABLED().key(),
+          "false",
+          SQLConf.AUTO_BROADCASTJOIN_THRESHOLD().key(),
+          "-1",
+          SparkSQLProperties.PRESERVE_DATA_GROUPING,
+          "true");
+
+  private static final Map<String, String> SPJ_OFF_SQL_CONF =
+      ImmutableMap.of(
+          SQLConf.V2_BUCKETING_ENABLED().key(),
+          "false",
+          SQLConf.REQUIRE_ALL_CLUSTER_KEYS_FOR_CO_PARTITION().key(),
+          "false",
+          SQLConf.ADAPTIVE_EXECUTION_ENABLED().key(),
+          "false",
+          SQLConf.AUTO_BROADCASTJOIN_THRESHOLD().key(),
+          "-1",
+          SparkSQLProperties.PRESERVE_DATA_GROUPING,
+          "true");
+
+  @BeforeClass
+  public static void setupSparkConf() {
+    spark.conf().set("spark.sql.shuffle.partitions", "4");
+  }
+
+  @After
+  public void removeTables() {
+    sql("DROP TABLE IF EXISTS %s", tableName);
+    sql("DROP TABLE IF EXISTS %s", tableName(OTHER_TABLE_NAME));
+  }
+
+  // TODO: add tests for truncate transforms once SPARK-40295 is released
+  // TODO: add tests for cases when one side contains a subset of keys once Spark supports this
+
```

Comment:
Should we have some tests with multiple buckets? or multiple partition transforms?

Reply:
I see the example below

Reply:
Does this work with a non Iceberg Source as part of the join?

Reply:
In theory it's possible. Spark will check if both sides of the SPJ have compatible partition transforms, via checking whether the V2 function identifiers are the same. So if the non-Iceberg source reports the same functions to Spark and use them in partition transforms, it could work (although I'm not sure whether this is a common use case).

Reply:
We will definitely need to support distributing the other side using the Iceberg function catalog. A common use case for this is MERGE. The incoming relation may be a view, which will never be distributed in a compatible way with Iceberg tables. I hope Spark would be smart enough to shuffle just the smaller relation using the reported partitioning from the target table.

Unfortunately, it does not work even if I distribute the incoming data manually using the Iceberg function catalog. We must have `KeyGroupPartitioning` on both sides, which can only be reported by data sources.

Reply:
At the moment, though, we can only support Iceberg to Iceberg joins.

Reply:
@sunchao, how hard will it be to add support for distributing the smaller relation using the reported partitioning by an Iceberg table for Spark 3.4?

Reply:
Yeah I really would want to be able to do something where the smaller side of the join is repartitioned using the transforms of the larger side if available.

Reply:
You are right if would only be beneficial if the keys are compatible on both sides. What about scenarios when I manually distribute the incoming relation using Iceberg functions? It still does not work right now as the other side is not `KeyGroupPartitioning`.

Reply:
What is the partitioning used by the other side?

Reply:
I'm mostly interested in
UnknownDF joined to Iceberg Table 
    If UnknownDF has no partitioning info and IcebergTable does, Spark requests a repartition of UnknownDF via KeyGroupPartitioning transforms, then does the SPJ

Reply:
For me, the question is how to repartition the UnknownDF.

On the Iceberg side, we have a set of N partitions with distinct keys. Now, we need to map the rows UnknownDF into these N partitions. However, at runtime, there could be some rows from UnknownDF that do not belong to any of these partitions.

Reply:
For inner-join, yes it is probably do-able, but it won't work for outer-joins.

Reply:
Does `MERGE INTO` only use inner-join? If so, we can start thinking about implementing this in Spark (however I'm not sure if it will make it into Spark 3.4). I briefly mentioned about the idea to the Spark community before, but they were concerned about adding `KeyGroupedPartitioning` to the shuffle mechanism. I feel there would be less objection if we introduce a flag to control the behavior though.

Reply:
It uses Inner Join for merges without a "when not matching" 

I would think for outer join we would similarly just keep all of the non-matching partitions separate but I haven't really looked at the code

Reply:
I'd be great to at least support cases when users manually repartition the incoming source relation using the Iceberg function catalog. I believe that would produce `HashPartitioning` or `RangePartitioning` on the other side.

Reply:
> @sunchao, we can't use the newly added logic for pushing down common keys cause we don't know the keys in the relation we will shuffle (until we actually shuffle it)?

Correct. We don't know the keys until runtime, which is too late.

> I'd be great to at least support cases when users manually repartition the incoming source relation using the Iceberg function catalog. I believe that would produce HashPartitioning or RangePartitioning on the other side.

Let me keep a note, and do some experiment on these ideas.

Reply:
Hi, I would like to apply this to streaming data source and have found this comments. I would like to use data from Kafka, then insert into a view and then use the SPJ to make the MERGE INTO cheaper.....  Right now the shuffle merge join is quite expensive. 

@sunchao, I know that your last comment was long time ago, but just wanted to know if you finally did some experiments.

Reply:
We've actually  made progress on the Spark side in https://github.com/apache/spark/pull/46255  , only a few corner case left like: https://github.com/apache/spark/pull/47064